### ISSUE TITLE
Are argmax / argmin with no axis argument not respecting the layout?

### ISSUE THREAD

**Issue by DavisVaughan** [QuantStack/xtensor-r#89]
It looks like the default layout (row major?) is always used when `argmax()` / `argmin()` are used without any `axis` arguments. I'd like to be able to be consistent with base R here if possible!
https://github.com/QuantStack/xtensor/blob/master/include/xtensor/xsort.hpp#L399

```cpp
// [[Rcpp::depends(xtensorrr)]]
// [[Rcpp::plugins(cpp14)]]

#include <xtensor-r/rarray.hpp>
#include <xtensor/xsort.hpp>
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
SEXP rray_argmax_cpp(xt::rarray<double> x) {
  // + 1 for R index consistency
  xt::rarray<int> res = xt::argmax(x) + 1;
  return res;
}
```

``` r
Rcpp::sourceCpp("~/Desktop/test.cpp")

# `4` is at position 3 (when starting from 1)
x_vec <- c(1, 3, 4, 2)

rray_argmax_cpp(x_vec)
#> [1] 3

x_mat <- matrix(x_vec, ncol = 2, byrow = TRUE)
x_mat
#>      [,1] [,2]
#> [1,]    1    3
#> [2,]    4    2

# 4 is at position 2 using R's column major layout
# (that's the right terminology, right?)
x_mat[2]
#> [1] 4

# This shows it at position 3!
rray_argmax_cpp(x_mat)
#> [1] 3
```

<sup>Created on 2019-01-09 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by DavisVaughan** [comment]
I can confirm that if I add 

```
#define XTENSOR_DEFAULT_LAYOUT ::xt::layout_type::column_major
```

to the top of the cpp file, I get the results I would expect!

**Comment by SylvainCorlay** [comment]
Thanks for reporting!

We could add a template argument to `argmax` to define the iteration order, and pass it to the `begin/end` iterator pair. So I think it is more a feature request.

Quick note: *I don't recommend changing `XTENSOR_DEFAULT_LAYOUT` to column_major event when you want to deal with column-major containers. The intent of the macro is for testing all test cases with an opposite convention (we could catch lots of issues with this). rarrays are already column-major anyways. If you use this macro in user code, you may find yourself binary incompatible with other libraries using the xtensor headers...*

**Comment by DavisVaughan** [comment]
A template arg would be appreciated!

I think I would also appreciate if keep_dims was passed through here.

Regarding your note, I had a feeling this was the case! It helps me with debugging too :)

**Comment by SylvainCorlay** [comment]
Excellent!

Do you feel like opening the PR to core xtensor? You would need to make add it as a *first* template argument defaulting to `layout_type::row_major`, and pass it through all the way to `begin - end` in https://github.com/QuantStack/xtensor/blob/6b385426db0b58ce496de596df274530aad48e38/include/xtensor/xsort.hpp#L399-L400

**Comment by DavisVaughan** [comment]
Sure!

Would it look something like

```cpp
template <layout_type L, class E, class F>
inline xtensor<std::size_t, 0> arg_func_impl(const E& e, F&& f)
{
  return cmp_idx(e.template begin<L>(),
                 e.template end<L>(), 1,
                 std::forward<F>(f));
}

template <layout_type L = layout_type::row_major, class E>
inline auto argmin(const xexpression<E>& e)
{
  using value_type = typename E::value_type;
  auto&& ed = eval(e.derived_cast());
  return detail::arg_func_impl<L>(ed, std::less<value_type>());
}
```

or would it somehow be a template argument and an argument to `arg_func_impl()`?

Should it be defaulted to `layout_type::row_major` or `XTENSOR_DEFAULT_LAYOUT`?

**Comment by SylvainCorlay** [comment]
That looks fine to me.

Initially, XTENSOR_DEFAULT_LAYOUT was only meant for layout, and not iteration order.

**Comment by DavisVaughan** [comment]
Closed by QuantStack/xtensor#1331

**Comment by SylvainCorlay** [comment]
:+1:
